include/envoy/buffer/buffer.h

```diff
@@ -440,5 +455,140 @@ class WatermarkFactory {
 
 using WatermarkFactoryPtr = std::unique_ptr<WatermarkFactory>;
 
+/**
+ * Holds an in-progress addition to a buffer.
+ *
+ * @note For performance reasons, this class is passed by value to
+ * avoid an extra allocation, so it cannot have any virtual methods.
+ */
+class Reservation final {
+public:
+  Reservation(Reservation&&) = default;
+
+  ~Reservation() {
+    if (!slices_.empty()) {
+      commit(0);
+    }
+  }
+
+  /**
+   * @return an array of `RawSlice` of length `numSlices()`.
+   */
+  RawSlice* slices() { return slices_.data(); }
+  const RawSlice* slices() const { return slices_.data(); }
+
+  /**
+   * @return the number of slices present.
+   */
+  uint64_t numSlices() const { return slices_.size(); }
+
+  /**
+   * @return the total length of the Reservation.
+   */
+  uint64_t length() const { return length_; }
+
+  /**
+   * Commits some or all of the data in the reservation.
+   * @param length supplies the number of bytes to commit. This must be
+   *   less than or equal to the size of the `Reservation`.
+   *
+   * @note No other methods should be called on the object after `commit()` is called.
+   */
+  void commit(uint64_t length) {
+    buffer_.commit(length, absl::MakeSpan(slices_), absl::MakeSpan(owned_slices_));
+    slices_.clear();
+    owned_slices_.clear();
+  }
+
+  // Tuned to allow reads of 128k, using 16k slices, and including one partially-used
+  // initial slice from a previous allocation.
+  static constexpr uint32_t NUM_ELEMENTS_ = 9;
```

Comment:
In the case where there is 1 byte available in a previous slice, 8 new slices of 16k will be allocated, plus the one previous slice, at least in the current implementation, and length will be set to 128k.

I think it would make more sense to return a reservation that includes the full length of the allocated slices, either (7 * 16k) + 1, or (8 * 16k) + 1. Right now it just returns (8 * 16k), and leaves (16k - 1) bytes unusable until the next read.

Reply:
Right.  So should we make this 8 instead of 9?

Reply:
Actually, that doesn't play well with the watermark limits. I need to think about this some more.